x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -332,6 +352,115 @@ private void createApiKeyAndIndexIt(
         }));
     }
 
+    public void updateApiKey(
+        final Authentication authentication,
+        final UpdateApiKeyRequest request,
+        final Set<RoleDescriptor> userRoles,
+        final ActionListener<UpdateApiKeyResponse> listener
+    ) {
+        ensureEnabled();
+
+        if (authentication == null) {
+            listener.onFailure(new IllegalArgumentException("authentication must be provided"));
+            return;
+        } else if (authentication.isApiKey()) {
+            listener.onFailure(new IllegalArgumentException("cannot use an api key to update api keys"));
+            return;
+        }
+
+        findVersionedApiKeyDocsForSubject(authentication, new String[] { request.getId() }, ActionListener.wrap((versionedDocs) -> {
+            final var apiKeyId = request.getId();
+
+            if (versionedDocs.isEmpty()) {
+                throw new ResourceNotFoundException("no api key owned by requesting user found for requested id [" + apiKeyId + "]");
+            }
+
+            validateCurrentApiKeyDocForUpdate(apiKeyId, authentication, single(versionedDocs).doc());
+
+            executeBulkIndexRequest(
```

Comment:
Following your suggestion, I've in-lined several other smaller methods. In terms of readability I have a slight preference to keep `executeBulkIndexRequest` though -- it keeps `updateApiKey` high level and essentially free of implementation details which I think is useful for someone just wanting a high level idea of what the method does. 

Let me know if you feel strongly!

Reply:
Thanks for your consideration. The comment was close to a "Nit". So I don't feel strongly about it.

Like you said, there is balance to strike. I am _not_ advocating for long-winded methods. Depending on the reader, exposing some of details may or may not be useful. As a code reviewer, I found them useful since I need to understand every single detail. That's probably true most of the time for people in our immediate team. But not so much for anyone else. 

It also has something to do with ES's general callback style coding. Most split'd methods make your read flow to jump from the innermost of a method to a different place. This coupled with the burdens of ensuring correct threadContext, error handling and code repitition decrease the value of one-off smaller methods.

Reply:
Absolutely see the benefit of having more things in one place. For now, I want to leave the current style and revisit when we get to implementing bulk updates since the overlap and method re-use there will further inform and possibly change the overall structure of the code.

Quite open to more in-lining at that stage. Thanks for the discussion around this!